### PR TITLE
Correctly handle duplicate solution folders

### LINKED ISSUE #413: Opening a solution with duplicate solution folder names crashes VS
Hi,

Since today I have a blocking issue loading my solutions in Visual Studio 2017.
After some research I've found out that by disabling the SonarLint extension for Visual Sudio 2017, my solutions are loading.

For now I'm going to use an older version so I can continue working.

### PR DESCRIPTION
**Pull request by duncanp-sonar**
Fixes #413

### CODE REVIEW

File: src/Integration.Vsix.UnitTests/Suppression/LiveIssueFactoryTests.cs [hunk 1]
```diff
@@ -59,31 +60,60 @@ public void Ctor_WithNullVsSolution_ThrowsArgumentNullException()
...
+        {
+            // See #413: https://github.com/SonarSource/sonarlint-visualstudio/issues/413
+
+            // Arrange
+            var vsSolutionMock = SetupSolutionMocks(
+                new KeyValuePair<string, string>("SolutionFolder1", "11111111-1111-1111-1111-111111111111"),
+                new KeyValuePair<string, string>("Item1", "22222222-2222-2222-2222-222222222222"),
+                new KeyValuePair<string, string>("SolutionFolder2", "33333333-3333-3333-3333-333333333333"),
+                new KeyValuePair<string, string>("Item1", "44444444-4444-4444-4444-444444444444"),
+                new KeyValuePair<string, string>("realProject.csproj", "55555555-5555-5555-5555-555555555555"),
```
Comment (duncanp-sonar) [inline src/Integration.Vsix.UnitTests/Suppression/LiveIssueFactoryTests.cs]: Real project files will return the fully-qualified path to the file, so there can't be any duplicates. However, solution folders are also treated as projects by VS, and the "path" that is returned is just the simple name. This means it's possible to have duplicates if there are nested folders e.g. _Folder1\ItemX_, _Folder2\ItemX_ - both will return a "path" of _ItemX_.

File: src/Integration.Vsix/Suppression/SuppressionManager.cs [hunk 2]
```diff
@@ -57,13 +56,21 @@ public SuppressionManager(IServiceProvider serviceProvider, ITimerFactory timerF
...
+                else
+                {
+                    CleanupSuppressionHandling();
+                }
             }
-            else
+            catch(Exception)
             {
-                CleanupSuppressionHandling();
+                // TODO: log exceptions to the output window
```
Comment (SonarTech) [inline src/Integration.Vsix/Suppression/SuppressionManager.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: src/Integration.Vsix/Suppression/LiveIssueFactory.cs [hunk 3]
```diff
@@ -121,25 +121,25 @@ public LiveIssue Create(SyntaxTree syntaxTree, Diagnostic diagnostic)
                 wholeLineText: lineText); // Line-level issue
         }
 
-        private void BuildProjectPathToIdMap(IVsSolution solution)
+        internal /* for testing */ static IDictionary<string, string> BuildProjectPathToIdMap(IVsSolution solution)
```
Comment (duncanp-sonar) [inline src/Integration.Vsix/Suppression/LiveIssueFactory.cs]: Changed to return the map to make testing easier.

File: src/Integration.Vsix/Suppression/LiveIssueFactory.cs [hunk 4]
```diff
@@ -121,25 +121,25 @@ public LiveIssue Create(SyntaxTree syntaxTree, Diagnostic diagnostic)
...
         {
-            projectPathToProjectIdMap = new Dictionary<string, string>(StringComparer.InvariantCultureIgnoreCase);
+            var map = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
```
Comment (duncanp-sonar) [inline src/Integration.Vsix/Suppression/LiveIssueFactory.cs]: Comparing file paths, so should be _ordinal_ rather _invariant_.

File: src/Integration.Vsix/Suppression/LiveIssueFactory.cs [hunk 5]
```diff
@@ -152,10 +152,12 @@ private void BuildProjectPathToIdMap(IVsSolution solution)
                 // to the project in the solution file.
                 // Fortunately, this is the logic used by soln5.GetGuidOfProjectFile... so we can just use that.
                 Guid projectGuid = soln5.GetGuidOfProjectFile(projectFile);
-                Debug.Assert(projectGuid != null, "Not expecting VS to return a null project guid");
```
Comment (duncanp-sonar) [inline src/Integration.Vsix/Suppression/LiveIssueFactory.cs]: Guids are value types.

**Comment by duncanp-sonar** [comment]
I've tested the fix in VS2015 and VS2017 with nested solution folders, and also with .Net Core projects (although .Net Core projects are treated exactly the same as other project types as far as these API calls are concerned).

**Comment by duncanp-sonar** [comment]
Excluding solution folders: not without calling additional APIs to check the project type. It didn't seem worth the extra effort to work out how, and it would add more code that would need to tested.

If we don't exclude the solution folders we'll end up with keys in the dictionary that will never be matched against, and in any normal solution the number of solution folders will be small (or at least much smaller than the number of real projects).

One option would be to exclude "fileNames" that aren't absolute paths. This would work the projects we've tested against currently, and unless the we have the absolute path the matching logic won't work anyway. I'll try that.

**Comment by Evangelink** [comment]
Not sure this is really worth it.

**Comment by duncanp-sonar** [comment]
No, probably not. I'll leave it as it is.

**Pull request merged**